File: rb/Tvmaze_sdk.rb
# Tvmaze SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'Tvmaze_types'


class TvmazeSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = TvmazeUtility.new
    @_utility = utility

    config = TvmazeConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features in the resolved order (make_options puts an explicit array
    # order first, else defaults to test-first). Ordering matters: the `test`
    # feature installs the base mock transport and the transport features
    # (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
    # must be added before them to sit at the base of the chain.
    feature_opts = TvmazeHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      featureorder = VoxgigStruct.getpath(@options, "__derived__.featureorder")
      if featureorder.is_a?(Array)
        featureorder.each do |fname|
          fopts = TvmazeHelpers.to_map(feature_opts[fname])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, TvmazeFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    TvmazeUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = TvmazeHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = TvmazeHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = TvmazeHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = TvmazeSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    # make_fetch_def returns a (fetchdef, err) tuple; destructure it and
    # return just the fetchdef Hash (raising on error) so callers — including
    # direct(), which indexes fetchdef["url"] — receive a Hash, mirroring the
    # ts/py prepare().
    fetchdef, fd_err = utility.make_fetch_def.call(ctx)
    raise fd_err if fd_err

    fetchdef
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue TvmazeError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = TvmazeHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = TvmazeHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Canonical facade: client.Aka.list / client.Aka.load({ "id" => ... })
  def Aka(data = nil)
    require_relative 'entity/aka_entity'
    AkaEntity.new(self, data)
  end


  # Canonical facade: client.AlternateList.list / client.AlternateList.load({ "id" => ... })
  def AlternateList(data = nil)
    require_relative 'entity/alternate_list_entity'
    AlternateListEntity.new(self, data)
  end


  # Canonical facade: client.Cast.list / client.Cast.load({ "id" => ... })
  def Cast(data = nil)
    require_relative 'entity/cast_entity'
    CastEntity.new(self, data)
  end


  # Canonical facade: client.CastCredit.list / client.CastCredit.load({ "id" => ... })
  def CastCredit(data = nil)
    require_relative 'entity/cast_credit_entity'
    CastCreditEntity.new(self, data)
  end


  # Canonical facade: client.CastMember.list / client.CastMember.load({ "id" => ... })
  def CastMember(data = nil)
    require_relative 'entity/cast_member_entity'
    CastMemberEntity.new(self, data)
  end


  # Canonical facade: client.Crew.list / client.Crew.load({ "id" => ... })
  def Crew(data = nil)
    require_relative 'entity/crew_entity'
    CrewEntity.new(self, data)
  end


  # Canonical facade: client.CrewCredit.list / client.CrewCredit.load({ "id" => ... })
  def CrewCredit(data = nil)
    require_relative 'entity/crew_credit_entity'
    CrewCreditEntity.new(self, data)
  end


  # Canonical facade: client.CrewMember.list / client.CrewMember.load({ "id" => ... })
  def CrewMember(data = nil)
    require_relative 'entity/crew_member_entity'
    CrewMemberEntity.new(self, data)
  end


  # Canonical facade: client.Episode.list / client.Episode.load({ "id" => ... })
  def Episode(data = nil)
    require_relative 'entity/episode_entity'
    EpisodeEntity.new(self, data)
  end


  # Canonical facade: client.GuestCastCredit.list / client.GuestCastCredit.load({ "id" => ... })
  def GuestCastCredit(data = nil)
    require_relative 'entity/guest_cast_credit_entity'
    GuestCastCreditEntity.new(self, data)
  end


  # Canonical facade: client.Image.list / client.Image.load({ "id" => ... })
  def Image(data = nil)
    require_relative 'entity/image_entity'
    ImageEntity.new(self, data)
  end


  # Canonical facade: client.Person.list / client.Person.load({ "id" => ... })
  def Person(data = nil)
    require_relative 'entity/person_entity'
    PersonEntity.new(self, data)
  end


  # Canonical facade: client.Schedule.list / client.Schedule.load({ "id" => ... })
  def Schedule(data = nil)
    require_relative 'entity/schedule_entity'
    ScheduleEntity.new(self, data)
  end


  # Canonical facade: client.ScheduledEpisode.list / client.ScheduledEpisode.load({ "id" => ... })
  def ScheduledEpisode(data = nil)
    require_relative 'entity/scheduled_episode_entity'
    ScheduledEpisodeEntity.new(self, data)
  end


  # Canonical facade: client.Search.list / client.Search.load({ "id" => ... })
  def Search(data = nil)
    require_relative 'entity/search_entity'
    SearchEntity.new(self, data)
  end


  # Canonical facade: client.Season.list / client.Season.load({ "id" => ... })
  def Season(data = nil)
    require_relative 'entity/season_entity'
    SeasonEntity.new(self, data)
  end


  # Canonical facade: client.Show.list / client.Show.load({ "id" => ... })
  def Show(data = nil)
    require_relative 'entity/show_entity'
    ShowEntity.new(self, data)
  end


  # Canonical facade: client.Update.list / client.Update.load({ "id" => ... })
  def Update(data = nil)
    require_relative 'entity/update_entity'
    UpdateEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = TvmazeSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: py/tvmaze_sdk.py
# Tvmaze SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import TvmazeUtility
from core.spec import TvmazeSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import TvmazeBaseFeature
from features import _make_feature


class TvmazeSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = TvmazeUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features in the resolved order (make_options puts an explicit
        # list order first, else defaults to test-first). Ordering matters: the
        # `test` feature installs the base mock transport and the transport
        # features (retry/cache/netsim/proxy/ratelimit) wrap whatever is
        # current, so `test` must be added before them to sit at the base.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            featureorder = vs.getpath(self.options, "__derived__.featureorder")
            if isinstance(featureorder, list):
                for fname in featureorder:
                    fopts = helpers.to_map(feature_opts.get(fname))
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return TvmazeUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = TvmazeSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def Aka(self, data=None) -> "AkaEntity":
        """Entity factory: client.Aka().list() / client.Aka().load({"id": ...})."""
        from entity.aka_entity import AkaEntity
        return AkaEntity(self, data)


    def AlternateList(self, data=None) -> "AlternateListEntity":
        """Entity factory: client.AlternateList().list() / client.AlternateList().load({"id": ...})."""
        from entity.alternate_list_entity import AlternateListEntity
        return AlternateListEntity(self, data)


    def Cast(self, data=None) -> "CastEntity":
        """Entity factory: client.Cast().list() / client.Cast().load({"id": ...})."""
        from entity.cast_entity import CastEntity
        return CastEntity(self, data)


    def CastCredit(self, data=None) -> "CastCreditEntity":
        """Entity factory: client.CastCredit().list() / client.CastCredit().load({"id": ...})."""
        from entity.cast_credit_entity import CastCreditEntity
        return CastCreditEntity(self, data)


    def CastMember(self, data=None) -> "CastMemberEntity":
        """Entity factory: client.CastMember().list() / client.CastMember().load({"id": ...})."""
        from entity.cast_member_entity import CastMemberEntity
        return CastMemberEntity(self, data)


    def Crew(self, data=None) -> "CrewEntity":
        """Entity factory: client.Crew().list() / client.Crew().load({"id": ...})."""
        from entity.crew_entity import CrewEntity
        return CrewEntity(self, data)


    def CrewCredit(self, data=None) -> "CrewCreditEntity":
        """Entity factory: client.CrewCredit().list() / client.CrewCredit().load({"id": ...})."""
        from entity.crew_credit_entity import CrewCreditEntity
        return CrewCreditEntity(self, data)


    def CrewMember(self, data=None) -> "CrewMemberEntity":
        """Entity factory: client.CrewMember().list() / client.CrewMember().load({"id": ...})."""
        from entity.crew_member_entity import CrewMemberEntity
        return CrewMemberEntity(self, data)


    def Episode(self, data=None) -> "EpisodeEntity":
        """Entity factory: client.Episode().list() / client.Episode().load({"id": ...})."""
        from entity.episode_entity import EpisodeEntity
        return EpisodeEntity(self, data)


    def GuestCastCredit(self, data=None) -> "GuestCastCreditEntity":
        """Entity factory: client.GuestCastCredit().list() / client.GuestCastCredit().load({"id": ...})."""
        from entity.guest_cast_credit_entity import GuestCastCreditEntity
        return GuestCastCreditEntity(self, data)


    def Image(self, data=None) -> "ImageEntity":
        """Entity factory: client.Image().list() / client.Image().load({"id": ...})."""
        from entity.image_entity import ImageEntity
        return ImageEntity(self, data)


    def Person(self, data=None) -> "PersonEntity":
        """Entity factory: client.Person().list() / client.Person().load({"id": ...})."""
        from entity.person_entity import PersonEntity
        return PersonEntity(self, data)


    def Schedule(self, data=None) -> "ScheduleEntity":
        """Entity factory: client.Schedule().list() / client.Schedule().load({"id": ...})."""
        from entity.schedule_entity import ScheduleEntity
        return ScheduleEntity(self, data)


    def ScheduledEpisode(self, data=None) -> "ScheduledEpisodeEntity":
        """Entity factory: client.ScheduledEpisode().list() / client.ScheduledEpisode().load({"id": ...})."""
        from entity.scheduled_episode_entity import ScheduledEpisodeEntity
        return ScheduledEpisodeEntity(self, data)


    def Search(self, data=None) -> "SearchEntity":
        """Entity factory: client.Search().list() / client.Search().load({"id": ...})."""
        from entity.search_entity import SearchEntity
        return SearchEntity(self, data)


    def Season(self, data=None) -> "SeasonEntity":
        """Entity factory: client.Season().list() / client.Season().load({"id": ...})."""
        from entity.season_entity import SeasonEntity
        return SeasonEntity(self, data)


    def Show(self, data=None) -> "ShowEntity":
        """Entity factory: client.Show().list() / client.Show().load({"id": ...})."""
        from entity.show_entity import ShowEntity
        return ShowEntity(self, data)


    def Update(self, data=None) -> "UpdateEntity":
        """Entity factory: client.Update().list() / client.Update().load({"id": ...})."""
        from entity.update_entity import UpdateEntity
        return UpdateEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "TvmazeSDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.aka_entity import AkaEntity
    from entity.alternate_list_entity import AlternateListEntity
    from entity.cast_entity import CastEntity
    from entity.cast_credit_entity import CastCreditEntity
    from entity.cast_member_entity import CastMemberEntity
    from entity.crew_entity import CrewEntity
    from entity.crew_credit_entity import CrewCreditEntity
    from entity.crew_member_entity import CrewMemberEntity
    from entity.episode_entity import EpisodeEntity
    from entity.guest_cast_credit_entity import GuestCastCreditEntity
    from entity.image_entity import ImageEntity
    from entity.person_entity import PersonEntity
    from entity.schedule_entity import ScheduleEntity
    from entity.scheduled_episode_entity import ScheduledEpisodeEntity
    from entity.search_entity import SearchEntity
    from entity.season_entity import SeasonEntity
    from entity.show_entity import ShowEntity
    from entity.update_entity import UpdateEntity
